tensorflow/compiler/xla/service/all_reduce_reassociate.cc

```diff
@@ -37,9 +40,57 @@ bool AreCompatible(const HloAllReduceInstruction *ar0,
   std::optional<AllReduceKey> key0 = GetAllReduceKey(ar0);
   std::optional<AllReduceKey> key1 = GetAllReduceKey(ar1);
   auto kind0 = MatchReductionComputation(ar0->to_apply());
+
   return key0 && key1 && kind0 && *key0 == *key1 && kind0 == op_kind;
 }
 
+bool AreCompatibleConverts(const HloInstruction *convert0,
+                           const HloInstruction *convert1) {
+  bool is_compatible = true;
+  // For numerical stability, we only re-order ops with casts from a narrow type
+  // to a wider type.
+  if (convert0) {
+    is_compatible &=
+        primitive_util::BitWidth(convert0->shape().element_type()) >
+        primitive_util::BitWidth(convert0->operand(0)->shape().element_type());
+  }
+
+  if (convert1) {
+    is_compatible &=
+        primitive_util::BitWidth(convert1->shape().element_type()) >
+        primitive_util::BitWidth(convert1->operand(0)->shape().element_type());
+  }
+
+  if (convert0 && convert1) {
+    is_compatible &=
+        convert0->shape().element_type() == convert1->shape().element_type() &&
```

Comment:
If both converts are present and they are inputs of an add/mul etc, then their shapes should be identical. So looks like this shape()->element_type() is redundant. May be convert it to an assert/CHECK and just check operand element types.

Reply:
changed it to a CHECK